modules/src/ics02_client/error.rs

```diff
@@ -1,120 +1,211 @@
+
+        Decode
+            [ TraceError<prost::DecodeError> ]
+            | _ | { "decode error" },
+
+        MissingHeight
+            | _ | { "invalid raw client consensus state: the height field is missing" },
+
+        InvalidClientIdentifier
+            [ ValidationError ]
+            | _ | { "invalid client identifier" },
+
+        InvalidRawHeader
+            [ TraceError<TendermintError> ]
+            | _ | { "invalid raw header" },
+
+        MissingRawHeader
+            | _ | { "missing raw header" },
+
+        DecodeRawMisbehaviour
+            [ TraceError<TendermintError> ]
+            | _ | { "invalid raw misbehaviour" },
+
+        InvalidRawMisbehaviour
+            [ ValidationError ]
+            | _ | { "invalid raw misbehaviour" },
+
+        MissingRawMisbehaviour
+            | _ | { "missing raw misbehaviour" },
+
+        InvalidHeightResult
+            | _ | { "height cannot end up zero or negative" },
+
+        InvalidAddress
+            | _ | { "invalid address" },
+
+        InvalidUpgradeClientProof
+            [ Ics23Error ]
+            | _ | { "invalid proof for the upgraded client state" },
+
+        InvalidUpgradeConsensusStateProof
+            [ Ics23Error ]
+            | _ | { "invalid proof for the upgraded consensus state" },
+
+        Tendermint
+            [ Ics07Error ]
+            | _ | { "tendermint error" },
+
+        InvalidPacketTimestamp
+            [ TraceError<TryFromIntError> ]
+            | _ | { "invalid packet timeout timestamp value" },
+
+        ClientArgsTypeMismatch
+            { client_type: ClientType }
+            | e | {
+                format_args!("mismatch between client and arguments types, expected: {0:?}",
+                    e.client_type)
+            },
+
+        RawClientAndConsensusStateTypesMismatch
+            {
+                state_type: ClientType,
+                consensus_type: ClientType,
+            }
+            | e | {
+                format_args!("mismatch in raw client consensus state {} with expected state {}",
+                    e.state_type, e.consensus_type)
+            },
+
+        LowHeaderHeight
+            {
+                header_height: Height,
+                latest_height: Height
+            }
+            | e | {
+                format!("received header height ({:?}) is lower than (or equal to) client latest height ({:?})",
+                    e.header_height, e.latest_height)
+            },
+
+        LowUpgradeHeight
+            {
+                upgraded_height: Height,
+                client_height: Height,
+            }
+            | e | {
+                format_args!("upgraded client height {0} must be at greater than current client height {1}",
+                    e.upgraded_height, e.client_height)
+            },
+    }
 }
 
-impl Kind {
-    pub fn context(self, source: impl Into<BoxError>) -> Context<Self> {
-        Context::new(self, Some(source.into()))
+impl Error {
```

Comment:
Do we need this?

Reply:
Yeah, I wondered about this as well. In the original code, it is not clear whether the error message is added for additional context, or because the implementer was not aware that the error source type is the same as the current error type. If we do not need that extra information when  displaying the error, then these functions can be removed.

Reply:
Why isn't it enough to add a block `InvalidRawClientState` in the define_error macro, but instead we need these manual methods?

Reply:
> If we do not need that extra information when displaying the error, then these functions can be removed.

The extra information does not seem particularly useful. We can get rid of it.

Reply:
> Why isn't it enough to add a block InvalidRawClientState in the define_error macro, but instead we need these manual methods?

Because the error source is the same as the error type defined. I have actually implemented recursive error definition in `flex-error` using the `[ Self ]` source, but I am a bit undecided on the usefulness. So far I think having recursive error source is ill-designed, and it is best to avoid. The only potential use I see is if the user want to add additional textual information in the error trace so that the user gets more context of what that particular error is related to.

> The extra information does not seem particularly useful. We can get rid of it.

Sounds good. I'll remove it then.